Seeds Driven Cluster Bootstrap

dlex opened the pull request:
## Cover letter

This PR addresses the second part of the scope of the ["Cluster Bootstrap" project](https://docs.google.com/document/d/1PaCQWgxTwh5f5v5ePkHFQKgOgic8mZpxkbyt8vM9LqY/edit#heading=h.axx8odyp4ut2). The first part was addressed by #6659.

This PR introduces a new node configuration parameter `empty_seed_starts_cluster` that can disable the _Empty Seed Cluster Bootstrap_ AKA _Root Driven Cluster Bootstrap_ (legacy) mode of bootstrapping a cluster, and allows the set of servers listed as seeds in each node configuration to start a cluster together, as soon as they form a raft group and elect a leader. The new bootstrapping mode is referred as _Seeds Driven Cluster Bootstrap_.

In either mode, cluster now gets cluster UUID reflected by a new controller log message, which lands second in the controller log right after the initial raft configuration message. Cluster UUID is also stored in kvstore of shard0 in every node.

In the new Seeds Driven bootstrap mode, all seed servers must be available for a cluster to be created, with identical node configurations. Afterwards, none of seed servers should try to form another new cluster ("split-brain") if their local storage is wiped out, unless all seed nodes are wiped together at the same time.

<!-- Describe in plain language the motivation (bug, feature, etc.) behind the change in this PR and how the included commits address it. -->

Fixes #333 

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [x] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

### Node Configuration in redpanda.yaml

* `empty_seed_starts_cluster` (default: true) to switch between the _Empty Seed Cluster Bootstrap_ (legacy) mode, and the _Seeds Driven Cluster Bootstrap_ mode. When disabled, it is required to be disabled in all seed nodes.
* `seed_servers` are required to be identical in every seed node when in the _Seeds Driven Cluster Bootstrap_ mode.

<!-- don't ship user breaking changes. Ping PMs for help with user visible changes  -->

## Release notes
<!--

If this PR does not need to be included in the release notes, then
simply have a bullet point for `none` directly under the `Release notes`
section, e.g.

* none

Otherwise, add one or more of the following sections. A section must have
at least 1 bullet point. You can add multiple sections with multiple
bullet points if this PR represents multiple release note items. See
the CONTRIBUTING.md guidelines for more details.
-->
### Features

* Seed driven cluster bootstrap mode. Disable `empty_seed_starts_cluster` to use it. That will allow the set of servers listed as seeds to start a cluster together. All seed servers must be available for a cluster to be created, with identical node configurations. Afterwards, none of seed servers will try to form another new cluster if their local storage is wiped out, unless all seed nodes are wiped together at the same time. Cluster now gets cluster UUID reflected by a new controller log message, and stored in kvstore.

<!-- * TODO: Short description of the feature. Explain how to configure the new feature if applicable. -->

### Improvements

<!-- * TODO: Short description of how this PR improves redpanda. -->
* Configurations of all nodes across the cluster can be identical
* Cluster is bootstrapped with all its seed servers
* Wiped out seed cluster members will not start their own new cluster

dlex commented:
Force-push update: 
* rebased onto [andrwng:redpanda/node_id](https://github.com/andrwng/redpanda/tree/node_id), the actual delivery branch for #6659
* integrated changes to `uuid_t`

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -66,13 +64,20 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
-    if (is_cluster_founder()) {
-        // TODO: we should only return non-empty seed list if our log is empty.
+    if (config::node().empty_seed_starts_cluster()) {
+        if (config::node().seed_servers().empty())
```
please use curly braces for all `if` statements even if they are one liners.

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -131,8 +136,39 @@ cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+        const config::seed_server& seed_server) {
+          return rpc_addr == seed_server.addr;
+      });
+    if (it == seed_servers.cend()) return {};
```
curly braces please

dlex replied:
np, I will check though all my code for that

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -131,8 +136,39 @@ cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+}
+
+/*static*/ std::optional<int32_t>
+cluster_discovery::get_node_index_in_seed_servers() {
```
should we include an information that we are looking for self id in the seed server list, not any node idx ?

dlex replied:
The info is here: https://github.com/dlex/redpanda/blob/333_cluster-bootstrap/src/v/cluster/cluster_discovery.cc#L73-L77

It will also log in case if `empty_seed_starts_cluster=true` and node_id is 0 for the root.

mmaslankaprv commented on src/v/cluster/security_manager.h:
```diff
@@ -42,8 +42,9 @@ class security_manager final {
 private:
-    template<typename Cmd, typename T>
-    ss::future<std::error_code> dispatch_updates_to_cores(Cmd, ss::sharded<T>&);
+    template<typename Cmd, typename Service>
+    ss::future<std::error_code>
+    dispatch_updates_to_cores(Cmd, ss::sharded<Service>&);
```
unrelated change, please extract this into another commit

mmaslankaprv commented on src/v/cluster/cluster_uuid.cc:
```diff
@@ -0,0 +1,38 @@
+const storage::kvstore::key_space cluster_uuid_key_space
+  = storage::kvstore::key_space::controller;
+
+std::optional<cluster_uuid> read_stored_cluster_uuid(storage::api& storage) {
+    std::optional<iobuf> cluster_uuid_buf = storage.kvs().get(
+      cluster_uuid_key_space, cluster_uuid_key);
+    if (cluster_uuid_buf) {
+        return cluster_uuid{
+          serde::from_iobuf<uuid_t>(std::move(*cluster_uuid_buf))};
+    }
+    return {};
+}
+
+ss::future<>
+write_stored_cluster_uuid(storage::api& storage, const cluster_uuid& value) {
+    return storage.kvs().put(
+      cluster_uuid_key_space, cluster_uuid_key, serde::to_iobuf(value));
+}
```
Should those method be part of `bootstrap_backend` ? Currently they have no context and can be executed from any shard while `bootstrap_backend` is living only on shard 0. This may lead to confusion as kvstore values are shard local

dlex replied:
Cluster UUID is read from kvstore before `bootstrap_backend` is started, otherwise I would have totally agreed with you. I'm thinking about exposing it from `bootstrap_backend` class as a static function though, so cluster UUID kvstore properties are encapsulated in `bootstrap_backend` and there is no need for the `cluster_uuid.cpp`.

However there is a different issue about kvstore being shard local, I'll chat today about it.

dlex replied:
This method is a part of `bootstrap_backend` now.

mmaslankaprv commented on src/v/cluster/types.h:
```diff
@@ -1939,6 +1940,20 @@ struct feature_update_license_update_cmd_data
+struct bootstrap_cluster_cmd_data
+  : serde::envelope<bootstrap_cluster_cmd_data, serde::version<0>> {
+    using rpc_adl_exempt = std::true_type;
+    static constexpr int8_t current_version = 0;
```
do we need version field ?

dlex replied:
I thought that's a requirement for serialization... or is it for ADL serialization only?

mmaslankaprv commented on src/v/cluster/bootstrap_backend.cc:
```diff
@@ -0,0 +1,53 @@
+                "Skipping bootstrap_cluster_cmd {}, current cluster_uuid: {}",
+                cmd.value.uuid,
+                *_cluster_uuid);
+              co_return errc::success;
```
do we want to inform caller that the id was already assigned with an error code ?

dlex replied:
I'm not sure if the caller can do anything meaningful about it other than loggging it, but this has already been logged here

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -69,7 +69,7 @@ std::vector<model::broker> cluster_discovery::initial_seed_brokers() const {
     if (config::node().empty_seed_starts_cluster()) {
         if (config::node().seed_servers().empty())
-        return {make_self_broker(config::node())};
+            return {make_self_broker(config::node())};
```
unrelated change

dlex replied:
that's a conflict resolution artifact, will fix it

mmaslankaprv commented on src/v/redpanda/application.cc:
```diff
@@ -1485,7 +1494,13 @@ void application::start_runtime_services(
+      ->start(
+        stored_cluster_uuid.has_value() ? std::vector<model::broker>{}
+                                        : cd.initial_seed_brokers(),
+        stored_cluster_uuid)
```
Can we make this logic, deciding about the initial servers a part of `cluster_discovery` ?

dlex replied:
It was there originally, and it looked like this:
https://github.com/andrwng/redpanda/blob/cluster_bootstrap/src/v/cluster/cluster_discovery.cc#L85-L88
https://github.com/andrwng/redpanda/blob/cluster_bootstrap/src/v/redpanda/application.cc#L1517

That seemed a bit awkward to me to pass an argument to a function only to return empty if it's true. So I've decided to take it out. After all, it depends on the perspective to the `cluster_discovery::intial_seed_brokers`: is it about geting seed brokers for unstarted cluster only, or about doing that _and_ also about embedding the knowledge whether there is a cluster into its result? The latter seemed to me a bit like a function doing more than one thing.

mmaslankaprv commented on src/v/cluster/controller.h:
```diff
@@ -127,6 +135,7 @@ class controller {
     friend controller_probe;
 
     ss::future<> cluster_creation_hook();
+
```
unrelated change

mmaslankaprv commented on src/v/cluster/controller.h:
```diff
@@ -120,11 +120,13 @@ class controller {
      */
     ss::future<> start(
-      std::vector<model::broker>,
+      std::vector<model::broker>&& initial_raft0_brokers,
```
why is this change required ?, if we need it we should make it a separate commit

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -26,6 +26,19 @@ using std::vector;
 
 namespace cluster {
 
+namespace {
+
+void verify_duplicate_seed_servers() {
+    std::vector<config::seed_server> s = config::node().seed_servers();
+    std::sort(s.begin(), s.end());
+    const auto s_dupe_i = std::adjacent_find(s.cbegin(), s.cend());
+    if (s_dupe_i != s.cend())
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format, "Duplicate items in seed_servers: {}", *s_dupe_i));
+}
+
+} // namespace
+
```
For me the main purpose of this commit is to validate the seed servers list. This is not reflected in the commit message

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -26,6 +26,19 @@ using std::vector;
 
+namespace {
+
+void verify_duplicate_seed_servers() {
```
maybe we can make this a part of configuration validation instead ?

mmaslankaprv commented on src/v/cluster/controller.cc:
```diff
@@ -427,39 +434,97 @@ ss::future<> controller::stop() {
+        vlog(clusterlog.trace, "Retrying to create cluster {}", cmd_data.uuid);
+    }
+}
+
```
Can we make this method part of members fronted or other service ?

mmaslankaprv commented on src/v/cluster/controller.cc:
```diff
@@ -427,39 +434,97 @@ ss::future<> controller::stop() {
+        // no feature is active, however the bootstrap_cluster_cmd is serde-only
+        const std::error_code errc = co_await _stm.invoke_on(
+          controller_stm_shard,
+          [this, cmd_data = std::move(cmd_data)](controller_stm& stm) mutable {
```
on retry `cmd_data` is accessed after move

mmaslankaprv commented on src/v/cluster/controller.cc:
```diff
@@ -436,8 +434,7 @@ ss::future<> controller::stop() {
     bootstrap_cluster_cmd_data cmd_data;
-    cmd_data.uuid = cluster_uuid(
-      uuid_t(boost::uuids::random_generator_mt19937()()));
+    cmd_data.uuid = cluster_uuid(uuid_t::create());
```
can we squash this change with one of the previous commits introducing this ?

mmaslankaprv commented:
Can you explaing what SDCB and ESCB mean ? Maybe adding this to cover letter would help

mmaslankaprv commented on src/v/cluster/tests/cluster_test_fixture.h:
```diff
@@ -114,9 +118,12 @@ class cluster_test_fixture {
+      configure_node_id use_node_id = configure_node_id::yes,
+      empty_seed_starts_cluster empty_seed_starts_cluster_val
+      = empty_seed_starts_cluster::yes) {
+        cluster::clusterlog.info("AWONG creating");
```
debug logging left overs

mmaslankaprv commented on src/v/cluster/cluster_discovery.h:
```diff
@@ -105,6 +108,11 @@ class cluster_discovery {
      */
     static std::optional<int32_t> get_node_index_in_seed_servers();
 
+    ss::future<std::vector<model::broker>> request_seed_brokers() const;
```
can we use `brokers` type here ?

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -151,6 +157,145 @@ cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+    });
+
+    bool failed = false;
+    std::vector<model::broker> seed_brokers;
```
use introduced `brokers` type

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -151,6 +157,145 @@ cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+         1s)]() mutable -> ss::future<ss::stop_iteration> {
+          result<cluster_bootstrap_info_reply> reply_result(
+            std::errc::connection_refused);
+          try {
```
maybe check abort source here before issuing a request

dlex commented:
force-push: rebased onto `dev`, addressed review feedback, made node_id assignment aware of results of cluster_bootstrap_info discovery, added support for the new command to log_viewer.

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -142,8 +148,41 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+cluster_discovery::get_node_index_in_seed_servers() {
+    const std::vector<config::seed_server>& seed_servers
+      = config::node().seed_servers();
+    vassert(
```
do we have a configuration validator that will error out gracefully before reaching this assertion ?

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -142,8 +148,41 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+        return {};
+    }
+    return {
+      boost::numeric_cast<int32_t>(std::distance(seed_servers.cbegin(), it))};
```
nit: we can use `model::node_id::type` here

mmaslankaprv commented on src/v/cluster/controller.cc:
```diff
@@ -454,39 +457,111 @@ ss::future<> controller::stop() {
+          model::timeout_clock::now() + 30s,
+          _as.local(),
+          std::nullopt);
+        if (errc == errc::success) co_return;
```
nit: curly braces

mmaslankaprv commented on src/v/cluster/controller.cc:
```diff
@@ -454,39 +457,111 @@ ss::future<> controller::stop() {
+          clusterlog.warn,
+          "Failed to replicate and wait the cluster bootstrap cmd, {}",
+          errc);
+        if (errc == errc::cluster_already_exists) co_return;
```
nit: braces

mmaslankaprv commented on src/v/cluster/bootstrap_backend.cc:
```diff
@@ -15,15 +15,53 @@
+
+std::error_code
+do_apply(user_and_credential cred, security::credential_store& store) {
+    if (store.contains(cred.username)) return errc::user_exists;
```
nit: braces

mmaslankaprv commented on src/v/cluster/bootstrap_backend.cc:
```diff
@@ -44,6 +82,17 @@ bootstrap_backend::apply_update(model::record_batch b) {
+              const std::error_code res
+                = co_await dispatch_updates_to_cores<user_and_credential>(
+                  *cmd.value.bootstrap_user_cred, _credentials);
+              if (res != errc::success) co_return res;
```
can be simplified with 
```    
        auto ec = dispatch...;
        if(ec) {
          co_return;
       }
```

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -69,22 +71,34 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
+    if (stored_cluster_uuid.has_value()) {
+        return {};
+    }
+    return initial_seed_brokers().get();
```
`get()` can not be called from generic seastar code (only inside of a seastar thread) This function should return future

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -148,18 +162,185 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+                    .then(&rpc::get_ctx_data<cluster_bootstrap_info_reply>);
+              });
+            if (reply_result) {
+                reply = std::move(reply_result.value());
```
can we `co_return reply.value()` here ? We wouldn't have to mutate reply local variable.

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -148,18 +162,185 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+}
+
+ss::future<> cluster_discovery::fetch_cluster_bootstrap_info() {
+    if (_cluster_bootstrap_info_fetched) co_return;
```
nit: curly braces

mmaslankaprv commented on src/v/cluster/cluster_discovery.h:
```diff
@@ -105,17 +135,36 @@ class cluster_discovery {
+     * \throw std::runtime_error if it cannot be either of them (validation has
+     * been failed)
+     */
+    bool is_amnesiac_seed_server() const;
```
nit: i like the name `amnesiac` but maybe it would be easier to grasp with `is_wiped_seed_server` ?

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -148,18 +162,185 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+
+ss::future<> cluster_discovery::fetch_cluster_bootstrap_info() {
+    if (_cluster_bootstrap_info_fetched) co_return;
+    const auto units = co_await ss::get_units(
```
i am not sure if i understand the need for mutual exclusion here, it seems that we control all entry points to the methods requesting cluster bootstrap state

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -148,18 +162,185 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+}
+
+void check_configuration_match(
+  bool& failure,
```
maybe it would be better to just return `bool` from this method

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -148,18 +162,185 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+        check_configuration_match(failed, peer.first, peer.second);
+        seed_brokers.push_back(std::move(peer.second.broker));
+    }
+    if (failed)
```
curly braces

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -69,22 +71,34 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
+  const std::optional<model::cluster_uuid>& stored_cluster_uuid) {
+    if (stored_cluster_uuid.has_value()) {
+        return {};
+    }
```
nit: rather than forcing the orchestration code to pass cluster uuid around, it seems like it would simplify things to have the stored cluster UUID as a member of `cluster_discovery`.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -185,4 +366,37 @@ cluster_discovery::get_node_index_in_seed_servers() {
+        }
+        check_configuration_match(failed, peer.first, peer.second);
+    }
+    if (failed)
```
nit: braces

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -185,4 +366,37 @@ cluster_discovery::get_node_index_in_seed_servers() {
+
+    bool failed = false;
+    std::optional<model::cluster_uuid> peer_cluster_uuid;
+    for (auto& peer : _cluster_bootstrap_info) {
```
nit: could use structured binding so it's easier to follow instead of `first` and `second`

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -185,4 +366,37 @@ cluster_discovery::get_node_index_in_seed_servers() {
+    std::optional<model::cluster_uuid> peer_cluster_uuid;
+    for (auto& peer : _cluster_bootstrap_info) {
+        if (peer.second.cluster_uuid) {
+            if (peer_cluster_uuid) {
```
nit: IMO easier to follow and reduces scoping written as:

```
if (peer_cluster_uuid == std::nullopt) {
    peer_cluster_uuid = peer.second.cluster_uuid;
    continue;
}
if (*peer_cluster_uuid != *peer.second.cluster_uuid) {
    ...
}

```

dlex replied:
```
if (peer_cluster_uuid == std::nullopt) {
    ...
    continue;
}
```
won't work, there's more in the loop body

andrwng replied:
Or how about something like:

```
for (const auto& [addr, reply] : _cluster_bootstrap_info) {
    check_configuration_match(failed, addr, reply);
    if (!reply.cluster_uuid) {
        continue;
    }
    if (peer_cluster_uuid == std::nullopt) {
        *peer_cluster_uuid = reply.cluster_uuid;
        continue;
    }
    if (*peer_cluster_uuid != reply.cluster_uuid) {
        ... error
        failed = true;
    }
}
```

Generally all the scopes make it a bit tricky to follow, and a bit error prone in that sense. IMO chipping away at "failure" cases with `continue` statements helps readability. Again though this is just cosmetic, so feel free to punt.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -185,4 +366,37 @@ cluster_discovery::get_node_index_in_seed_servers() {
       boost::numeric_cast<int32_t>(std::distance(seed_servers.cbegin(), it))};
 }
 
+bool cluster_discovery::is_amnesiac_seed_server() const {
```
nit: maybe more obvious what this means if called something like `seeds_already_have_cluster_uuid` or `remote_cluster_uuid_exists` or somesuch

It seems possible to get here by configuration only, without "wiping out" a node. E.g. start a new node configured as a seed, but expect it to join an existing cluster

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -69,22 +71,34 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
-        return {};
+        co_return brokers{};
     }
     if (get_node_index_in_seed_servers()) {
```
nit: IMO this is awkward to reuse here, because we're not using the returned value `idx`. It makes me wonder if we can replace it with some `is_configured_seed_server()` method that we can maybe cache once. Or maybe we should keep the seed servers around as a set and use some `seed_servers.contains(self)` style of checking

dlex replied:
Well `idx` is not used, however the complexy to obtain `idx` is the same as to obtain the flag whether we are a part of `seed_servers` at all. Caching that value is possible yet I don't think we can gain anything visible from it.

andrwng replied:
My motivation is to make the code easier to understand upon first glance. IMO it's very obvious what it means for this node to be configured as a seed server -- it's less obvious what it means to have `get_node_index..()` return `nullopt`.

You're right, there is not functional change though, so feel free to punt.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -148,18 +162,185 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+        peers.emplace_back(seed_server.addr, cluster_bootstrap_info_reply{});
+    }
+    co_await ss::parallel_for_each(peers, [this](auto& peer) -> ss::future<> {
+        peer.second = co_await request_cluster_bootstrap_info_single(
+          peer.first);
```
Are we guaranteed that this returns a reply, e.g. if the seed is down?

Just want to make sure we're able to get through here when at least one seed is down.

dlex replied:
Actually it won't, and your point is correct, this shouldn't be the case for wiped seed nodes. I will attend to this in a subsequent PR.

andrwng commented on src/v/redpanda/application.cc:
```diff
@@ -1496,6 +1496,27 @@ void application::start_runtime_services(
+    // Initialize the Raft RPC endpoint before the rest of the runtime RPC
+    // services so the cluster seeds can elect a leader and write a cluster
+    // UUID before proceeding with the rest of bootstrap.
+    const bool start_raft_rpc_early = !stored_cluster_uuid.has_value();
```
I think this should also be tied to cluster discovery via some `is_founding_cluster()` call or somesuch -- non-seeds probably don't need to do this.

dlex replied:
Changing to
```
    const bool start_raft_rpc_early = !stored_cluster_uuid.has_value()
                                      && cd.is_cluster_founder().get();
```

andrwng commented on src/v/cluster/controller.cc:
```diff
@@ -446,39 +467,113 @@ ss::future<> controller::stop() {
+    for (simple_time_jitter<model::timeout_clock> retry_jitter(1s);;) {
+        model::record_batch b = serde_serialize_cmd(
+          bootstrap_cluster_cmd{0, cmd_data});
+        // cluster::replicate_and_wait() cannot be used here because it checks
+        // for serde_raft_0 feature to be enabled. Before a cluster is here,
+        // no feature is active, however the bootstrap_cluster_cmd is serde-only
```
nit: update comment

dlex replied:
It is up to date, `cluster::replicate_and_wait()` is avoided here for exactly that reason.

andrwng replied:
Ah oops, missed that this was the `cluster::` variant.

dlex commented on src/v/cluster/controller.cc:
```diff
@@ -454,39 +457,111 @@ ss::future<> controller::stop() {
+
+    for (simple_time_jitter<model::timeout_clock> retry_jitter(1s);;) {
+        model::record_batch b = serde_serialize_cmd(
+          bootstrap_cluster_cmd{0, cmd_data});
```
Great point, I'll attend to that in a subsequent PR.

dlex commented on src/v/cluster/controller.cc:
```diff
@@ -454,39 +457,111 @@ ss::future<> controller::stop() {
+      _raft0->last_visible_index() > model::offset{}
+      || _raft0->config().brokers().size() > 1) {
+        // The controller log has already been written to
+        // TODO: create_cluster() but use the metrics' cluster_id
```
We have discussed before that it's convenient to have the identical values for cluster ID and cluster UUID, and for that purpose the latter should be initialized from the former in the cluster upgrade scenario.

andrwng replied:
I think we're still not really on the same page here yet. I was under the impression we would never initialize a `cluster_uuid` with anything other than a random value. Reconciliation of `cluster_id` and `cluster_uuid` would take the form of: in metrics only, if we have a `cluster_id`, use it, if not, use the `cluster_uuid`, where it is possible that `cluster_id` and `cluster_uuid` are not identical.

We have no guarantee that `cluster_id` is even a UUID, since it can be set by the user.

dlex commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -142,8 +162,241 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+    bool failed = false;
+    brokers seed_brokers;
+    seed_brokers.reserve(_cluster_bootstrap_info.size());
+    for (auto& peer : _cluster_bootstrap_info) {
```
Great catch!

dlex commented:
force-push: rebase conflicts resolved, most review comments addressed, DT test filure fixed

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -142,8 +202,259 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+    // identical configuration. Note that the order is preserved from
+    // seed_servers to seed_brokers, but the latter is missing one item for
+    // self.
+    const std::vector<config::seed_server>& self_seed_servers
+      = config::node().seed_servers();
+    for (auto [ib, is]
+         = std::tuple{seed_brokers.begin(), self_seed_servers.cbegin()};
+         ib != seed_brokers.cend();
+         ++ib) {
+        while (is != self_seed_servers.cend()
+               && is->addr != ib->rpc_address()) {
+            ++is;
+        }
+        if (ib->id() == model::unassigned_node_id) {
+            vassert(
+              is != self_seed_servers.cend(),
+              "seed_brokers and seed_servers mismatch");
+            const auto idx = boost::numeric_cast<model::node_id::type>(
+              std::distance(self_seed_servers.cbegin(), is));
+            ib->replace_unassigned_node_id(model::node_id{idx});
+        }
+    }
```
nit: I'm having a really hard time understanding this code. Does this work instead?
```
(instead of populating seed_brokers above, do it here?)
model::node_id::type idx = 0;
brokers seed_brokers;
for (const auto& ss : self_seed_servers) {
    auto& broker = _cluster_bootstrap_info[ss.addr].broker;
    if (broker.id == model::unassigned_node_id) {
        broker.replace_unassigned_node_id(model::node_id(idx));
    }
    seed_brokers.emplace_back(std::move(broker));
    ++idx;
}
```

mmaslankaprv replied:
i would follow andrew suggestion here, seems like a major simplification

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -142,8 +202,259 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+    if (_cluster_bootstrap_info_fetched) {
+        co_return;
+    }
+    const auto units = co_await _fetch_cluster_bootstrap_mx.get_units(_as);
+    if (_cluster_bootstrap_info_fetched) {
+        co_return;
+    }
```
nit: this is surprising to see because we only ever call this in series. I think having this here will give readers the wrong impression that they should expect parallel calls, unless this is something you intend on adding immediately.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -71,14 +71,74 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
+        co_return brokers{};
     }
-    return {};
+    if (get_node_index_in_seed_servers()) {
```
Should this actually be `is_cluster_founder()`? Then we'd even be able to avoid the duplicate call to `fetch_cluster_bootstrap_info()`.

Also it handles the case when another node has a cluster UUID (i.e. the local node is not a cluster founder).

Generally, it'd be great if instead of smearing various calls to figure out whether we're a founder around the bootstrap path (e.g. `stored_cluster_uuid`), we should figure up ASAP whether or not we should be founding a cluster. To that end, we make all the calls to storage and the bootstrap endpoints up front, and figure out whether a cluster UUID exists locally, or among the seed servers. From then on, we can keep that cached in `cluster_discovery` and have `cluster_discovery` handle behavioral changes based on whether we're a founder.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -71,14 +71,74 @@ ss::future<node_id> cluster_discovery::determine_node_id() {
+    if (get_node_index_in_seed_servers()) {
+        co_await fetch_cluster_bootstrap_info();
+        brokers seed_brokers = get_seed_brokers_for_bootstrap();
+        seed_brokers.push_back(make_self_broker(config::node()));
```
nit: why not just have the call to `get_seed_brokers_for_bootstrap()` do this?

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -142,8 +202,259 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+        peer.second = co_await request_cluster_bootstrap_info_single(
+          peer.first);
+        co_return;
+    });
```
nit: it seems we have spread some of the validations across different calls in `cluster_discovery`. I think it would make reading through this much easier if all validations happened up front. Is that feasible, such that by the end of this function we know:
- whether we should be founding a cluster (by the absence of a cluster UUID)
- if there are cluster UUIDs, that they all match
- that all seeds' configurations match if we _are_ a founder
- that all seeds' versions match if we _are_ a founder

Then we can remove some code from `get_seed_brokers_for_bootstrap` and `remote_cluster_uuid_exists`. At least for me, as I walk through the code, it's really hard to keep track of what has been validated and what hasn't, and I think doing a bunch of checks up front would help. It also helps us fail fast.

andrwng commented on src/v/cluster/bootstrap_service.cc:
```diff
@@ -10,14 +10,33 @@
+      std::back_inserter(r.seed_servers),
+      [](const config::seed_server& seed_server) { return seed_server.addr; });
+    r.empty_seed_starts_cluster = config::node().empty_seed_starts_cluster();
+    r.cluster_uuid = co_await cluster::read_stored_cluster_uuid_front(_storage);
```
nit: IMO a cleaner alternative here is to bake `cluster_uuid` into `storage::api` that gets set by `bootstrap_backend`, and have users of the `cluster_uuid` access it through storage. Also it seems odd to hit disk more than once -- why not cache in `storage::api`?

andrwng reviewed (commented):
Looking better, mostly just cosmetic feedback so far.

mmaslankaprv commented on src/v/cluster/bootstrap_backend.h:
```diff
@@ -0,0 +1,58 @@
+class bootstrap_backend final {
+public:
+    bootstrap_backend(
+      const std::optional<model::cluster_uuid>&, ss::sharded<storage::api>&);
```
i remember we were chatting about making reading `cluster_uuid` part of bootstrap backend, is it possible to do this ?

dlex replied:
I have redesigned this part, the current `cluster_uuid` is now in `storage::api` like `node_uuid` and is accessed by `bootstrap_backend` and other classes from there.

mmaslankaprv commented on tests/rptest/services/redpanda.py:
```diff
@@ -585,6 +585,12 @@ def __init__(self,
         # stash a copy here so that we can quickly look up e.g. addresses later.
         self._node_configs = {}
 
+        self._seed_servers = [self.nodes[0]]
```
is this intended ? it seems that it cause some tests to fail

dlex replied:
yes. The template for that is:
```
{% if include_seed_servers %}
  seed_servers:
  {% for node in seed_servers %}
    - host:
        address: {{node.account.hostname}}
        port: 33145
  {% endfor %}
{% endif %}
```
and `include_seed_servers` is initialized as
```
        include_seed_servers = True
        node_id = self.idx(node)
        if omit_seeds_on_idx_one and node_id == 1:
            include_seed_servers = False
```
so the quoted line will make `seed_servers` a root node everywhere except for the root node (for root-driven bootstrap)

dlex replied:
actually as @andrwng has pointed out, this will make `PartitionBalancerTest.test_full_nodes` fail because that test creates a faux `Redpanda` instanse to skip actual testing in debug. I'm posting a fix.

dlex commented:
/ci-repeat 3

jcsp commented on src/v/cluster/feature_manager.cc:
```diff
@@ -153,6 +153,10 @@ ss::future<> feature_manager::start() {
           [this] { return maybe_log_license_check_info(); });
     });
 
+    for (const model::node_id n : seed_nodes) {
```
I think this makes sense on first startup, but not on subsequent upgrades.

When we upgrade, we should not fast-forward all nodes to the latest version: they have to come up and report their version as normal, otherwise we might start sending new-format messages to old-version nodes.

dlex replied:
`seed_nodes` is only populated when we are starting up a new cluster. It contains either 
* the root node on the root node in the old way of cluster bootstrap (root-driven), or 
* all seed severs on a seed server, and all other seed servers have been verified to be at the same logical version as the local (in seeds-driven bootstrap).

I think I can name this parameter better to reflect that.

dlex commented:
force push: review comments addressed, `cluster_uuid` handling refactored (stored in `storage::api`, i/o done only once where needed). A bug preventing a wiped seed to rejoin fixed.

dlex commented:
force push: fix a linter error

mmaslankaprv commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -182,18 +196,259 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
     if (config::node().empty_seed_starts_cluster()) {
         if (config::node().seed_servers().empty()) {
-            return node_id{0};
+            co_return node_id{0};
```
do we validate that in this case node_id configured in `redpanda.yml` matches the  one returned from here ?

dlex replied:
nope, I agree that it should be added

dotnwat commented on src/v/cluster/feature_manager.cc:
```diff
@@ -73,7 +73,8 @@ feature_manager::feature_manager(
 
-ss::future<> feature_manager::start() {
+ss::future<>
+feature_manager::start(std::vector<model::node_id>&& cluster_founder_nodes) {
```
please take by value
